javav2/example_code/s3/src/main/java/com/example/s3/express/S3DirectoriesScenario.java

```diff
@@ -0,0 +1,578 @@
+        long normalTimeStart = System.nanoTime();
+        for (int index = 0; index < downloads; index++) {
+            if (index % 10 == 0) {
+                logger.info("Download " + index + " of " + downloads);
+            }
+
+            try {
+                s3DirectoriesActions.getObjectAsync(mS3RegularClient, mregularBucketName, bucketObject).join();
+            } catch (CompletionException ce) {
+                Throwable cause = ce.getCause();
+                if (cause instanceof S3Exception) {
+                    logger.error("S3Exception occurred: {}", cause.getMessage(), ce);
+                } else {
+                    logger.error("An unexpected error occurred: {}", cause.getMessage(), ce);
+                }
+            }
+        }
+
+        long normalTimeDifference = System.nanoTime() - normalTimeStart;
+        logger.info("The directory bucket took " + directoryTimeDifference + " nanoseconds, while the normal bucket took " + normalTimeDifference + " nanoseconds.");
+        long difference = normalTimeDifference - directoryTimeDifference;
+        logger.info("That's a difference of " + difference + " nanoseconds, or");
+        logger.info(difference / 1_000_000_000.0 + " seconds.");
+
+        if (difference < 0) {
+            logger.info("The directory buckets were slower. This can happen if you are not running on the cloud within a VPC.");
+        }
+        waitForInputToContinue(scanner);
+    }
+
+    private static String createSessionAddObjects() {
+        logger.info("""    
+            5. Create an object and copy it over.
+            We'll create a basic object consisting of some text and upload it to the 
+            normal bucket. 
+            Next we'll copy the object into the Directory bucket using the regular client. 
+            This works fine because copy operations are not restricted for Directory buckets.
+            """);
+        waitForInputToContinue(scanner);
+
+        String bucketObject = "basic-text-object.txt";
+        try {
+            s3DirectoriesActions.putObjectAsync(mS3RegularClient, mregularBucketName, bucketObject, "Look Ma, I'm a bucket!").join();
+            s3DirectoriesActions.createSessionAsync(mS3ExpressClient, mdirectoryBucketName).join();
+
+            // Copy the object to the destination S3 bucket.
+            s3DirectoriesActions.copyObjectAsync(mS3ExpressClient, mregularBucketName, bucketObject, mdirectoryBucketName, bucketObject).join();
+        } catch (CompletionException ce) {
+            Throwable cause = ce.getCause();
+            if (cause instanceof S3Exception) {
+                logger.error("S3Exception occurred: {}", cause.getMessage(), ce);
+            } else {
+                logger.error("An unexpected error occurred: {}", cause.getMessage(), ce);
+            }
+        }
+        logger.info(""" 
+            It worked! It's important to remember the user permissions when interacting with 
```

Comment:
Why exactly did it work? Is this because the S3Client that performed the copy operation is the expressClient using the credentials for the user with permission to work with directory buckets?

Reply:
In java - the copyObjectAsync worked, In Python - the copy_object methof worked. Once again - following the approved program.